src/storage/mvcc/reader.rs

```diff
@@ -390,18 +399,30 @@ impl MvccReader {
                         lock_valid = false;
                     }
                 }
-                match (w_key, l_key) {
+                match (w_user_key.clone(), l_key) {
                     (None, None) => return Ok(None),
                     (None, Some(k)) => Key::from_encoded(k.to_vec()),
-                    (Some(k), None) => Key::from_encoded(k.to_vec()).truncate_ts()?,
-                    (Some(wk), Some(lk)) => if wk < lk {
+                    (Some(k), None) => k,
+                    (Some(wk), Some(lk)) => if wk.encoded().as_slice() < lk {
                         Key::from_encoded(lk.to_vec())
                     } else {
-                        Key::from_encoded(wk.to_vec()).truncate_ts()?
+                        wk
                     },
                 }
             };
-            if let Some(v) = self.get(&key, ts)? {
+
+            if !write_valid
+                || (desc && w_user_key.is_some()
+                    && w_user_key.as_ref().unwrap().encoded().as_slice()
```

Comment:
It seems `w_user_key.as_ref().unwrap().encoded().as_slice()
                        != key.encoded().as_slice()` is true for L407, could we do something to avoid this complex `compare`?